library/std/src/sys/unix/fs.rs

```diff
@@ -1648,55 +1667,155 @@ mod remove_dir_impl {
+    fn unlink_direntry(ent: &DirEntry, parent_fd: RawFd) -> io::Result<bool> {
+        match is_dir(&ent) {
+            Some(true) => Ok(false),
+            Some(false) => {
+                cvt(unsafe { unlinkat(parent_fd, ent.name_cstr().as_ptr(), 0) })?;
+                Ok(true)
+            }
+            None => {
+                match cvt(unsafe { unlinkat(parent_fd, ent.name_cstr().as_ptr(), 0) }) {
                     // type unknown - try to unlink
                     Err(err)
                         if err.raw_os_error() == Some(libc::EISDIR)
                             || err.raw_os_error() == Some(libc::EPERM) =>
                     {
-                        // if the file is a directory unlink fails with EISDIR on Linux and EPERM everyhwere else
-                        remove_dir_all_recursive(Some(fd), Path::new(&child.file_name()))?;
+                        // if the file is a directory unlink fails with EISDIR on Linux
+                        // and EPERM everyhwere else
+                        Ok(false)
                     }
-                    result => {
-                        result?;
-                    }
-                },
+                    result => result.map(|_| true),
+                }
             }
         }
+    }
 
-        // unlink the directory after removing its contents
-        cvt(unsafe {
-            unlinkat(parent_fd.unwrap_or(libc::AT_FDCWD), pcstr.as_ptr(), libc::AT_REMOVEDIR)
+    struct CachedReadDir {
+        readdir: ReadDir,
+        raw_fd: RawFd,
+    }
+
+    struct DirComponent {
+        name: CString,
+        ino: u64,
+    }
+
+    fn readdir_open_path(path: &CStr) -> io::Result<(DirComponent, CachedReadDir)> {
+        let dir_fd = openat_nofollow_dironly(None, path)?;
+
+        // use fstat() to get the inode of the directory
+        let mut stat = unsafe { mem::zeroed() };
+        cvt(unsafe { fstat64(dir_fd.as_raw_fd(), &mut stat) })?;
+        let (readdir, raw_fd) = fdreaddir(dir_fd)?;
+        Ok((
+            DirComponent { name: CString::new("")?, ino: stat.st_ino },
+            CachedReadDir { readdir, raw_fd },
+        ))
+    }
+
+    fn readdir_open_child(
+        readdir: &CachedReadDir,
+        child: &DirEntry,
+    ) -> io::Result<(DirComponent, CachedReadDir)> {
+        let dir_fd = openat_nofollow_dironly(Some(readdir.raw_fd), child.name_cstr())?;
+        let (readdir, raw_fd) = fdreaddir(dir_fd)?;
+        Ok((
+            DirComponent { name: child.name_cstr().into(), ino: child.entry.d_ino },
+            CachedReadDir { readdir, raw_fd },
+        ))
+    }
+
+    fn readdir_reopen_parent(
+        dir: &CachedReadDir,
+        expected_parent_dir: &DirComponent,
+    ) -> io::Result<CachedReadDir> {
+        let parent_dir_fd = openat_nofollow_dironly(Some(dir.raw_fd), unsafe {
+            CStr::from_bytes_with_nul_unchecked(b"..\0")
         })?;
```

Comment:
I'm not sure about the semantics, but opening the parent like this seems very risky...

A fool proof but slow solution is keeping a file descriptor for the first directory, and whenever you run out of file descriptors and can't go back to parent, rewind that first directory stream and go again.

Reply:
Making this safe is what the long [discussion above](https://github.com/rust-lang/rust/pull/93160#discussion_r789702060) is about.

Reply:
Ok, I tried reading through it, will have to try again.

Given that this function has once again had trouble with distrusted directories, I would advocate for the simplest solution (keep parent fd, rewind back and start from the beginning again) as a first step, even if it's absurdly slower...

Studying code from coreutils from multiple OS's might also be a good bet (instead of only the strace).

GNU Coreutils, for example, uses the FTS interface to go through the directories recursively.

Reply:
> GNU Coreutils, for example, uses the FTS interface to go through the directories recursively.

If explore this route, we would have to use `FTS_NOCHDIR` at the very least. We can't change directories for library use, while `rm` is free to do so.

Reply:
I know that coreutils rm uses fts under the hood. But fts is not universal (e.g. musl does not have it), not standardized via POSIX and horribly broken on some platforms, e.g. on Macos where `rm -r dir` on a deep directory hierarchy results in `rm: fts_read: File name too long` errors. I haven't tested BSD. I think rolling our own on top of `openat()`, `fdreaddir()`, ... is the best choice. I have an implementation using iterated `openat()` lying around which we can use if we deem going up via .. to dangerous.

Reply:
Looking into this more coreutils uses [gnulib fts](https://github.com/coreutils/gnulib/blob/master/lib/fts.c) in the proprietary `FTS_CWDFD` mode. The check for parent dir dev/inode is [here](https://github.com/coreutils/gnulib/blob/master/lib/fts.c#L2045-L2068).